File: frontend/js/utils/validator.js
/**
 * 表单验证工具模块
 */
const Validator = {
    // 验证规则
    rules: {
        required: (value) => value !== '' && value !== null && value !== undefined,
        email: (value) => /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(value),
        phone: (value) => /^1[3-9]\d{9}$/.test(value),
        minLength: (value, len) => value && value.length >= len,
        maxLength: (value, len) => !value || value.length <= len,
        min: (value, minVal) => !isNaN(value) && Number(value) >= minVal,
        max: (value, maxVal) => !isNaN(value) && Number(value) <= maxVal,
        pattern: (value, regex) => regex.test(value),
        url: (value) => /^https?:\/\/.+/.test(value),
        number: (value) => !isNaN(value) && value !== '',
        integer: (value) => /^\d+$/.test(value)
    },

    // 错误消息
    messages: {
        required: '此字段为必填项',
        email: '请输入有效的邮箱地址',
        phone: '请输入有效的手机号码',
        minLength: '最少需要{0}个字符',
        maxLength: '最多允许{0}个字符',
        min: '最小值为{0}',
        max: '最大值为{0}',
        pattern: '格式不正确',
        url: '请输入有效的URL地址',
        number: '请输入数字',
        integer: '请输入整数'
    },

    /**
     * 验证单个字段
     * @param {*} value - 字段值
     * {Array} rules - 验证规则
     * @returns {Object}
     */
    validateField(value, rules) {
        for (const rule of rules) {
            let ruleName, ruleParam;

            if (typeof rule === 'string') {
                ruleName = rule;
                ruleParam = null;
            } else if (typeof rule === 'object') {
                ruleName = rule.name;
                ruleParam = rule.param;
            } else {
                continue;
            }

            const validateFunc = this.rules[ruleName];
            if (!validateFunc) continue;

            let isValid;
            try {
                isValid = ruleParam !== null
                    ? validateFunc(value, ruleParam)
                    : validateFunc(value);
            } catch (error) {
                isValid = false;
            }

            if (!isValid) {
                let message = this.messages[ruleName] || '验证失败';
                if (ruleParam !== null) {
                    message = message.replace('{0}', ruleParam);
                }
                return { valid: false, message };
            }
        }
        return { valid: true };
    },

    /**
     * 验证整个表单
     * @param {Object} formData - 表单数据
     * @param {Object} schema - 验证规则
     * @returns {Object}
     */
    validateForm(formData, schema) {
        const errors = {};
        let isValid = true;

        for (const [field, rules] of Object.entries(schema)) {
            const value = formData[field];
            const result = this.validateField(value, rules);

            if (!result.valid) {
                errors[field] = result.message;
                isValid = false;
            }
        }

        return { valid: isValid, errors };
    },

    /**
     * 清理输入，防止XSS
     * @param {string} text - 输入文本
     * @returns {string}
     */
    sanitize(text) {
        if (!text) return text;

        const div = document.createElement('div');
        div.textContent = text;
        return div.innerHTML;
    },

    /**
     * 清理对象中的字符串值
     * @param {Object} obj - 输入对象
     * @returns {Object}
     */
    sanitizeObject(obj) {
        const result = {};
        for (const [key, value] of Object.entries(obj)) {
            if (typeof value === 'string') {
                result[key] = this.sanitize(value);
            } else if (typeof value === 'object' && value !== null) {
                result[key] = this.sanitizeObject(value);
            } else {
                result[key] = value;
            }
        }
        return result;
    }
};

// 全局可用
window.Validator = Validator;

File: backend/utils/validator.py
"""
数据验证工具
"""
import re
from typing import Dict, List, Any, Optional
from datetime import datetime


class Validator:
    """数据验证工具类"""

    # 验证规则
    rules = {
        'required': lambda value: value is not None and value != '',
        'email': lambda value: re.match(r'^[^\s@]+@[^\s@]+\.[^\s@]+$', str(value)) is not None,
        'phone': lambda value: re.match(r'^1[3-9]\d{9}$', str(value)) is not None,
        'min_length': lambda value, min_len: len(str(value)) >= min_len,
        'max_length': lambda value, max_len: len(str(value)) <= max_len,
        'min_value': lambda value, min_val: float(value) >= min_val,
        'max_value': lambda value, max_val: float(value) <= max_val,
        'pattern': lambda value, regex: re.match(regex, str(value)) is not None,
        'integer': lambda value: isinstance(value, int) or str(value).isdigit(),
        'positive': lambda value: float(value) > 0,
        'date': lambda value: Validator._is_valid_date(value),
    }

    # 错误消息
    messages = {
        'required': '此字段为必填项',
        'email': '请输入有效的邮箱地址',
        'phone': '请输入有效的手机号码',
        'min_length': '最少需要{0}个字符',
        'max_length': '最多允许{0}个字符',
        'min_value': '最小值为{0}',
        'max_value': '最大值为{0}',
        'pattern': '格式不正确',
        'integer': '必须是整数',
        'positive': '必须是正数',
        'date': '日期格式不正确，请使用YYYY-MM-DD格式',
    }

    @staticmethod
    def _is_valid_date(value: str) -> bool:
        """验证日期格式"""
        try:
            datetime.strptime(str(value), '%Y-%m-%d')
            return True
        except (ValueError, TypeError):
            return False

    @classmethod
    def validate_field(cls, value: Any, rules: List) -> Dict[str, Any]:
        """
        验证单个字段

        Args:
            value: 字段值
            rules: 验证规则列表

        Returns:
            验证结果
        """
        for rule in rules:
            if isinstance(rule, str):
                rule_name = rule
                rule_param = None
            elif isinstance(rule, dict):
                rule_name = rule.get('name')
                rule_param = rule.get('param')
            else:
                continue

            # 获取验证函数
            validate_func = cls.rules.get(rule_name)
            if not validate_func:
                continue

            # 执行验证
            try:
                if rule_param is not None:
                    is_valid = validate_func(value, rule_param)
                else:
                    is_valid = validate_func(value)

                if not is_valid:
                    message = cls.messages.get(rule_name, '验证失败')
                    if rule_param is not None:
                        message = message.replace('{0}', str(rule_param))
                    return {'valid': False, 'message': message}
            except Exception:
                return {'valid': False, 'message': '验证过程出错'}

        return {'valid': True}

    @classmethod
    def validate_form(cls, data: Dict[str, Any], schema: Dict[str, List]) -> Dict[str, Any]:
        """
        验证整个表单

        Args:
            data: 表单数据
            schema: 验证规则

        Returns:
            验证结果
        """
        errors = {}
        is_valid = True

        for field, rules in schema.items():
            value = data.get(field)
            result = cls.validate_field(value, rules)
            if not result['valid']:
                errors[field] = result['message']
                is_valid = False

        return {'valid': is_valid, 'errors': errors}

    @staticmethod
    def sanitize_string(value: str) -> str:
        """
        清理字符串，防止XSS

        Args:
            value: 输入字符串

        Returns:
            清理后的字符串
        """
        if not value:
            return value

        # 转义特殊字符
        value = value.replace('&', '&amp;')
        value = value.replace('<', '&lt;')
        value = value.replace('>', '&gt;')
        value = value.replace('"', '&quot;')
        value = value.replace("'", '&#x27;')

        return value

    @staticmethod
    def sanitize_dict(data: Dict[str, Any]) -> Dict[str, Any]:
        """
        清理字典中的字符串值

        Args:
            data: 输入字典

        Returns:
            清理后的字典
        """
        result = {}
        for key, value in data.items():
            if isinstance(value, str):
                result[key] = Validator.sanitize_string(value)
            elif isinstance(value, dict):
                result[key] = Validator.sanitize_dict(value)
            else:
                result[key] = value
        return result
